pkg/convert/dynamic.go

```diff
@@ -0,0 +1,183 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package convert
+
+import (
+	"fmt"
+	"math/big"
+
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+)
+
+// To support DynamicPseudoType, perform a best-effort conversion at the data level without any type information.
+func newDynamicEncoder() Encoder {
+	return &dynamicEncoder{}
+}
+
+type dynamicEncoder struct{}
+
+// Convert a PropertyValue to a tftypes.Value. If the optional ty parameter is specified, ensure that the resulting
+// tftypes.Value is constructed with the given type.
+func (enc *dynamicEncoder) fromPropertyValue(p resource.PropertyValue) (tftypes.Value, error) {
+	switch {
+	case propertyValueIsUnknown(p):
+		return tftypes.NewValue(tftypes.Object{}, tftypes.UnknownValue), nil
+	case p.IsNull():
+		return tftypes.NewValue(tftypes.Object{}, nil), nil
+	case p.IsBool():
+		return tftypes.NewValue(tftypes.Bool, p.BoolValue()), nil
+	case p.IsNumber():
+		return tftypes.NewValue(tftypes.Number, p.NumberValue()), nil
+	case p.IsString():
+		return tftypes.NewValue(tftypes.String, p.StringValue()), nil
+	case p.IsArray():
+		var result []tftypes.Value
+		for _, e := range p.ArrayValue() {
+			te, err := enc.fromPropertyValue(e)
+			if err != nil {
+				return tftypes.Value{}, err
+			}
+			result = append(result, te)
+		}
+		// tftypes.NewValue is pretty strict in that array elements must have the same type.
+		// Try to encode this as an array and fallback on a tuple if the types do not match.
+		if commonTy, err := tftypes.TypeFromElements(result); err == nil {
+			return tftypes.NewValue(tftypes.List{ElementType: commonTy}, result), nil
+		} else {
+			types := []tftypes.Type{}
+			for _, r := range result {
+				types = append(types, r.Type())
+			}
+			return tftypes.NewValue(tftypes.Tuple{ElementTypes: types}, result), nil
+		}
+	case p.IsAsset():
+		return tftypes.Value{}, fmt.Errorf("Assets inside dynamically typed blocks are not yet supported")
+	case p.IsArchive():
+		return tftypes.Value{}, fmt.Errorf("Archives inside dynamically typed blocks are not yet supported")
+	case p.IsSecret() || p.IsOutput() && p.OutputValue().Secret:
+		return tftypes.Value{}, fmt.Errorf("Secrets inside dynamically typed blocks are not yet supported")
+	case p.IsOutput():
+		// Cannot be unknown at this point, or secret. Strip dependencies and send the value.
```

Comment:
Instead of a comment here, let's just `assert` it.
```suggestion
		o := p.OutputValue()
		contract.Assertf(o.Known && !o.Secret,
		  "Cannot be unknown at this point, or secret.")
```